src/backend/oneapi/kernel/approx.hpp

```diff
@@ -0,0 +1,619 @@
+                   sycl::accessor<Ty> d_in, KParam in, int ioff, Tp x,
+                   int method, int batch, bool doclamp, int batch_dim) const {
+        const int grid_x = floor(x);    // nearest grid
+        const Tp off_x   = x - grid_x;  // fractional offset
+
+        const int x_lim    = in.dims[XDIM_];
+        const int x_stride = in.strides[XDIM_];
+        const int idx      = ioff + grid_x * x_stride;
+
+        bool cond[4] = {grid_x - 1 >= 0, true, grid_x + 1 < x_lim,
+                        grid_x + 2 < x_lim};
+        int off[4]   = {cond[0] ? -1 : 0, 0, cond[2] ? 1 : 0,
+                      cond[3] ? 2 : (cond[2] ? 1 : 0)};
+
+        Ty zero = (Ty)0;
+
+        for (int n = 0; n < batch; n++) {
+            Ty val[4];
+            int idx_n = idx + n * in.strides[batch_dim];
+            for (int i = 0; i < 4; i++) {
+                val[i] = (doclamp || cond[i]) ? d_in[idx_n + off[i] * x_stride]
+                                              : zero;
+            }
+            bool spline = method == AF_INTERP_CUBIC_SPLINE;
+            d_out[ooff + n * out.strides[batch_dim]] =
+                cubicInterpFunc(val, off_x, spline);
+        }
+    }
+
+   private:
+    sycl::accessor<Ty> d_yo_;
+    const KParam yo_;
+    sycl::accessor<Ty> d_yi_;
+    const KParam yi_;
+    sycl::accessor<Tp> d_xo_;
+    const KParam xo_;
+    const Tp xi_beg_;
+    const Tp xi_step_reproc_;
+    const Ty offGrid_;
+    const int blocksMatX_;
+    const int batch_;
+    const int method_;
+    const int XDIM_;
+    const int INTERP_ORDER_;
+};
+
+template<typename Ty, typename Tp>
+void approx1(Param<Ty> yo, const Param<Ty> yi, const Param<Tp> xo,
+             const int xdim, const Tp xi_beg, const Tp xi_step,
+             const float offGrid, const af_interp_type method,
+             const int order) {
+    constexpr int THREADS = 256;
+
+    auto local         = sycl::range{THREADS, 1};
+    dim_t blocksPerMat = divup(yo.info.dims[0], local[0]);
+    auto global        = sycl::range{blocksPerMat * local[0] * yo.info.dims[1],
+                              yo.info.dims[2] * yo.info.dims[3] * local[1]};
+
+    // Passing bools to opencl kernels is not allowed
+    bool batch =
+        !(xo.info.dims[1] == 1 && xo.info.dims[2] == 1 && xo.info.dims[3] == 1);
+
+    getQueue().submit([&](sycl::handler &h) {
+        auto yoAcc = yo.data->get_access(h);
+        auto yiAcc = yi.data->get_access(h);
+        auto xoAcc = xo.data->get_access(h);
+        sycl::stream debugStream(128, 128, h);
+
+        h.parallel_for(
+            sycl::nd_range{global, local},
+            approx1Kernel<Ty, Tp>(yoAcc, yo.info, yiAcc, yi.info, xoAcc,
+                                  xo.info, xi_beg, Tp(1) / xi_step, (Ty)offGrid,
+                                  (int)blocksPerMat, (int)batch, (int)method,
+                                  xdim, order));
+    });
+    ONEAPI_DEBUG_FINISH(getQueue());
+}
+
```

Comment:
If possible, can we split to two files? approx1 and approx2?

Reply:
done